src/substrait/function_registry.py

```diff
@@ -0,0 +1,316 @@
+                parameter_name
+            ].SerializeToString(deterministic=True) != dtype.SerializeToString(
+                deterministic=True
+            ):
+                return False
+            else:
+                parameters[parameter_name] = dtype
+                return True
+
+    kind = dtype.WhichOneof("kind")
+
+    if kind != expected_kind:
+        return False
+
+    if kind == "decimal":
+        if violates_integer_option(
+            dtype.decimal.scale, parameterized_type.decimal.scale, parameters
+        ) or violates_integer_option(
+            dtype.decimal.precision, parameterized_type.decimal.precision, parameters
+        ):
+            return False
+
+    # TODO handle all types
+
+    return True
+
+
+class FunctionEntry:
+    def __init__(self, uri: str, name: str, impl: Mapping[str, Any]) -> None:
+        self.name = name
+        self.normalized_inputs: list = []
+        self.uri: str = uri
+        self.anchor = next(id_generator)
+        self.arguments = []
+        self.rtn = impl["return"]
+        self.nullability = impl.get("nullability", False)
+        self.variadic = impl.get("variadic", False)
+        if input_args := impl.get("args", []):
+            for val in input_args:
+                if typ := val.get("value"):
+                    self.arguments.append(to_parameterized_type(typ.strip("?")))
+                    self.normalized_inputs.append(normalize_substrait_type_names(typ))
+                elif arg_name := val.get("name", None):
+                    self.arguments.append(val.get("options"))
+                    self.normalized_inputs.append("req")
+
+    def __repr__(self) -> str:
+        return f"{self.name}:{'_'.join(self.normalized_inputs)}"
+
+    def satisfies_signature(self, signature: tuple) -> Optional[str]:
+        if self.variadic:
+            min_args_allowed = self.variadic.get("min", 0)
+            if len(signature) < min_args_allowed:
+                return None
+            inputs = [self.arguments[0]] * len(signature)
+        else:
+            inputs = self.arguments
+        if len(inputs) != len(signature):
+            return None
+
+        zipped_args = list(zip(inputs, signature))
+
+        parameters = {}
+
+        for x, y in zipped_args:
+            if type(y) == str:
+                if y not in x:
+                    return None
+            else:
+                if not covers(y, x, parameters):
+                    return None
+
+        return evaluate(self.rtn, parameters)
+
+
+class FunctionRegistry:
+    def __init__(self) -> None:
+        self._function_mapping: dict = defaultdict(dict)
+        self.id_generator = itertools.count(1)
+
+        self.uri_aliases = {}
+
+        for fpath in importlib_files("substrait.extensions").glob(  # type: ignore
+            "functions*.yaml"
+        ):
+            uri = f"https://github.com/substrait-io/substrait/blob/main/extensions/{fpath.name}"
```

Comment:
If someone has their own set of extension files they would be unable to use this routine.

Reply:
This is only meant to preload extensions from substrait repo by default for convenience, they would still be able to ignore these and call `register_extension_yaml` method for their own extension files instead. Do you mean that they would waste resources by loading default extensions if they know they won't use it? I guess we can make it configurable in the constructor:

```
def __init__(self, load_default_extensions: bool = True)
```

Reply:
If someone had their own directory of YAML files and wanted to use this routine they would instead have to do their own directory visiting sequence.  Not a big deal but if we start going down the dialect path more (as discussed at the community meeting) this may be more common in the future (which means this is fine for now).

Reply:
Understood. The reason I went with `register_extension_yaml` and `register_extension_dict` only was because I needed them for default extensions and tests, respectively. Beyond that, it's a bit unclear right now whether people will have a local directory of yamls available or maybe just a collection of http paths, so I decided to leave it up to the users. Having said that, we can always add other helper methods to make popular cases convenient, of course.